moveit_core/trajectory_processing/test/test_cartesian_speed.cpp

```diff
@@ -0,0 +1,143 @@
+void printTrajectory(robot_trajectory::RobotTrajectory& trajectory)
+{
+  const robot_model::JointModelGroup* group = trajectory.getGroup();
+  const std::vector<int>& idx = group->getVariableIndexList();
+  unsigned int count = trajectory.getWayPointCount();
+
+  ROS_INFO_STREAM_NAMED("trajectory_processing",
+                        "trajectory length is " << trajectory.getWayPointDurationFromStart(count - 1) << " seconds.");
+  for (unsigned i = 0; i < count; i++)
+  {
+    robot_state::RobotStatePtr point = trajectory.getWayPointPtr(i);
+    ROS_INFO_STREAM_NAMED("trajectory_processing", "Waypoint " << i << " time "
+                                                               << trajectory.getWayPointDurationFromStart(i) << " pos "
+                                                               << point->getVariablePosition(idx[0]) << " vel "
+                                                               << point->getVariableVelocity(idx[0]) << " acc "
+                                                               << point->getVariableAcceleration(idx[0]));
+
+    if (i > 0)
+    {
+      robot_state::RobotStatePtr prev = trajectory.getWayPointPtr(i - 1);
+      ROS_INFO_STREAM_NAMED(
+          "trajectory_processing",
+          "jrk " << (point->getVariableAcceleration(idx[0]) - prev->getVariableAcceleration(idx[0])) /
+                        (trajectory.getWayPointDurationFromStart(i) - trajectory.getWayPointDurationFromStart(i - 1)));
+    }
+  }
+}
+
+TEST(TestCartesianSpeed, TestCartesianEndEffectorSpeed)
+{
+  trajectory_processing::IterativeParabolicTimeParameterization time_parameterization;
+  EXPECT_EQ(initStraightTrajectory(TRAJECTORY), 0);
+
+  EXPECT_TRUE(time_parameterization.computeTimeStamps(TRAJECTORY));
+  trajectory_processing::setMaxCartesianEndEffectorSpeed(TRAJECTORY, 0.01);
+  printTrajectory(TRAJECTORY);
+  size_t num_waypoints = TRAJECTORY.getWayPointCount();
+  robot_state::RobotStatePtr kinematic_state = TRAJECTORY.getFirstWayPointPtr();
+  Eigen::Isometry3d current_end_effector_state = kinematic_state->getGlobalLinkTransform("r_wrist_roll_link");
+  Eigen::Isometry3d next_end_effector_state;
+  double euclidean_distance = 0.0;
+
+  for (size_t i = 0; i < num_waypoints - 1; i++)
+  {
+    kinematic_state = TRAJECTORY.getWayPointPtr(i + 1);
+    next_end_effector_state = kinematic_state->getGlobalLinkTransform("r_wrist_roll_link");
+    euclidean_distance +=
+        pow(pow(next_end_effector_state.translation()[0] - current_end_effector_state.translation()[0], 2) +
+                pow(next_end_effector_state.translation()[1] - current_end_effector_state.translation()[1], 2) +
+                pow(next_end_effector_state.translation()[2] - current_end_effector_state.translation()[2], 2),
+            0.5);
+    current_end_effector_state = next_end_effector_state;
+  }
+  double avg_speed = euclidean_distance / TRAJECTORY.getWayPointDurationFromStart(num_waypoints);
+  ASSERT_DOUBLE_EQ(0.01, avg_speed);
+}
```

Comment:
Hi, 
I do not quite understand what you mean by hand calculated. If I would calculate the distance by hand I would use the exact same formula since it is the mathematically correct formula. 
So can you please specify a bit more what you mean by hand calculated? 
Thanks

Reply:
Maybe the meaning is to hard-code a reference trajectory to test the overall output? For instance with 3-4 trajectory points? Would this make sense?

Reply:
> Maybe the meaning is to hard-code a reference trajectory to test the overall output? For instance with 3-4 trajectory points? Would this make sense?

This is closer to what I meant. Comparing the output from the function against predetermined expected values is good practice.

Reply:
Ok sounds good. Then I will define some points in cartesian space and plan the cartesian trajectory and then it is easy to compare. 
The problem I am facing is that I don't know how to load the IK solver in the test script so that I can use plan in cartesian space. I am not familiar enough with the Moveit code base for that. Can you give me a hint where to find that information?